rnnalyse/extractors/base_extractor.py

```diff
@@ -147,38 +146,73 @@ def _print_time_info(self, start_time: float, print_every: int) -> None:
         """ Generates the embeddings of a sentence and writes to file.
 
         Parameters
         ----------
         sentence : Sentence
             The to-be-extracted sentence, represented as a list of strings.
+        selection_func: Callable
+            Function which determines if activations for a token should be extracted or not.
+
+        Returns
+        -------
+        num_extracted: int
+            Number of extracted activations for this sentence.
         """
-        def _incremental_avg(old_avg, new_value):
-            # n_sens only get incremented after _extract_sentence is called, therefore + 1 here
-            return old_avg + 1 / (self.n_sens + 1) * (new_value - old_avg)
 
-        sen_activations: PartialActivationDict = self._init_sen_activations(len(sentence))
+        sen_activations: PartialActivationDict = self._init_sen_activations(0)
 
         activations: FullActivationDict = self.init_lstm_states.states
 
-        for i, token in enumerate(sentence):
+        for i, token in enumerate(sentence.sen):
             out, activations = self.model(token, activations)
 
-            for l, name in self.activation_names:
-                sen_activations[(l, name)][i] = activations[l][name].detach().numpy()
+            # Check whether current activations match criterion defined in selection_func
+            if selection_func(i, token, sentence):
+                for layer, name in self.activation_names:
+                    sen_activations[(layer, name)] = np.append(
+                        sen_activations[(layer, name)], activations[layer][name].detach().numpy()[np.newaxis, ...],
+                        axis=0
+                    )
 
-        # Update average eos states
-        self.avg_eos_states = {
-            l: {
-                'hx': _incremental_avg(self.avg_eos_states[l]['hx'], activations[l]['hx']),
-                'cx': _incremental_avg(self.avg_eos_states[l]['cx'], activations[l]['cx'])
-            }
-            for l in self.avg_eos_states
+        return sen_activations
+
+    def extract_average_eos_activations(self):
+        """ Extract average end of sentence activations and dump them to a file. """
+
+        def _incremental_avg(old_avg, new_value, n_sens):
+            return old_avg + 1 / n_sens * (new_value - old_avg)
+
+        def _eos_selection_func(pos, token, sentence):
+            return pos == len(sentence.sen) - 1
+
+        # Init states
+        avg_eos_states = {
+            layer: {'hx': torch.zeros(self.model.hidden_size), 'cx': torch.zeros(self.model.hidden_size)}
+            for layer in range(self.model.num_layers)
         }
 
-        self._dump_activations(sen_activations)
+        # Extract
+        for i, labeled_sentence in enumerate(self.corpus.values()):
+            sen_activations = self._extract_sentence(labeled_sentence, _eos_selection_func)
```

Comment:
These are already the eos activations, right? not the sentence activations? Maybe reflect that in the name of the variable (eos_activations = ...)?

Reply:
Done!